requests-for-comments/2023-03-01-Temporal-RFC.md

```diff
@@ -0,0 +1,239 @@
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Nice to pick up where you left off
+- **DAG-able** - Most ETLs resemble a Directed Acyclical Graph. Move A → B → C
+    - You can think of this as **Composable**
+- **Parallelize-able** - We should be able to run multiple bits of the DAG at once
+- **Timeouts** - We should be able to set timeouts on the DAG and Nodes
+- **Cancelable** - We should be able to cancel the DAG’s execution *gracefully*
+- **Terminate-able** - We should be able to rip the carpet out from under the DAG
+- **Nonblocking failures** - A failure of one DAG should not halt the execution of others.
+
+### Webhooks
+
+Ahh webhooks. Webhooks are a particularly sticky problem.
+
+Requirements:
+
+- **Idempotent** or **exactly once** - no duplicates
+- **Retry-able** - If the thing fails, get back on that horse
+- **Resumable** - Helps with idempotency/exactly once
+- **Parallelize-able** - We should be able to run multiple webhooks at once and out of order (non-blocking)
+- **Timeouts** - We should be able to set timeouts for webhooks
+- **Terminate-able** - We should be able to rip the carpet out from under webhooks
+- **Nonblocking failures** - A failure of one webhook endpoint should not halt the execution of others.
+
+## Requirements Grid
+
+| Task | Idempotent / Exactly Once | Retry-able | Resumable | Parallelize-able | Timeouts | Cancelable | Terminate-able | Nonblocking Failures |
+| --- | --- | --- | --- | --- | --- | --- | --- | --- |
+| Billing | 💯 | 💯 | 💯 | 🙏 | 🙏 | 🙏 | 💯 | 💯 |
+| Celery Task |  | 💯 | 🙏 | 💯 | 💯 |  | 💯 | 💯 |
+| ETLs | 💯 | 💯 | 💯 | 💯 | 💯 | 💯 | 💯 | 💯 |
+| Webhooks | 💯 | 💯 | 🙏 | 💯 | 🙏 |  | 💯 | 💯 |
+
+🙏 - Really nice to have
+
+💯 - Required
+
+# Where are we at now?
+
+Currently our plugin server has been doing all of the heavy lifting and has gotten us quite far in the **************2 years************** that it has existed. I once gave feedback that the plugin server is a diamond encrusted hammer that tempts you into thinking that everything is a nail. Joke was on me because it turned out that it got us way further than I thought.
+
+Since this began we’ve done amazing work in keeping the plugin server going. Chief among the improvements is that we split the plugin server up dependent on the workload into pools of resources that are responsible for different categories of plugin.
+
+### The plugin server, a survey
+
+So what all is the plugin server responsible for? There are really two types of plugins or tasks that we operate on the plugin server: Stateful and Stateless. Stateful plugins make side effects when they are called. This means that an API is called or a mutation is made in a db. Stateless plugins only affect the event as it is in flight. A good example of a stateless plugin is GeoIP where we take the ip that is in the event and based on that information and a static GeoIP db we geolocalize using the ip. 
```

Comment:
Ah I'd misunderstood what stateless and stateful plugins were. I thought it was a reference to if it was updating the [Node.js VM context](https://nodejs.org/api/vm.html#vm-executing-javascript) with team specific data such that you can't share a VM.

Will have to go and read the code again to understand properly 🔍

Reply:
Agreed on `Stateful and Stateless` being a bit confusing to me too, what you're describing is closer to the `pure / impure` function concept?

Reply:
What Harry is saying is indeed how we treat it in plugin server, which is probably not what it means in functional programming /rest of the world.

For plugins, having a config => stateful, even though it might not have side effects. So: ex: advanced-geoip is stateful, when all it does is what geoIP does, but with an extra config option which decides which events to add info to.

Reply:
Sorry I should have been more clear here because these are terms that we use in the Plugin server. I am referring to the Functional context of stateful and stateless. My original draft actually used side-effecty and side effect less, but I swapped that out. 

I just found it a bit easier to categorize things if I though about our plugins as either being pure or side-effecty functions. The pure functions are much easier to reason about from a performance perspective, the side-effecty ones are not so easy and we need to consider many other ways that they can fail considering they have external dependencies.

Reply:
Makes sense 👍